core/templates/span.h

```diff
@@ -0,0 +1,106 @@
+/**************************************************************************/
+/*  span.h                                                                */
+/**************************************************************************/
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef SPAN_H
+#define SPAN_H
+
+#include "core/typedefs.h"
+
+// Equivalent of std::span.
+// Represents a view into a contiguous memory space.
+template <typename T>
+class Span {
+	T *_ptr = nullptr;
+	size_t _len = 0;
+
+public:
+	Span() = default;
+	Span(T *p_ptr, size_t p_len) :
+			_ptr(p_ptr), _len(p_len) {}
+
+	T *ptrw() { return _ptr; }
```

Comment:
Actually no, it wouldn't, not directly, as vieww() would already call _ref(). I think then that we need to make default, copy, and move assignment and constructors private of this thing so it can't escape scope. I have some concerns about these things ending up outliving insertions or removals from cowdata.

Reply:
It doesn't, but it _will_ behave unexpectedly if trying to access a `ptrw` from a const view:

```
Span<const int> span = _cowdata.view();
const int *ptr = span.ptrw();  // Since T is const int
```

I think I better use SFINAE to disable `ptrw` for `is_const_v<T>` types.

Reply:
It's a bit complex, but I think this is better: 
```c++
std::enable_if<std::negation_v<std::is_const<T>>, std::remove_const_t<T>> *ptrw() { return _ptr; }
```

Reply:
That would be good I think. Too bad there's not really a way to ensure that something like 

```
y = x.view()
x.insert(blah)
y.find("something")
```

can't happen. Or worse that y gets passed to something else and stored.

Reply:
Yeah, that's going to be an inherent risk with this data type. Would be possible to avoid in Rust, but not good old C++

Reply:
Yeah, makes sense. I'll add a comment about the SFINAE use. I will also add a disclaimer to not store this data type.

Reply:
might help to delete copy constructors and such, with a comment as to why. at least it will catch some cases of likely abuse.

Reply:
Okay, let's not do that then. It's no more dangerous than sending a ptr() or ptrw() anyway, so maybe it's not actually a problem